### PR TITLE
fixed memcache and add new memcached driver

### PR DESCRIPTION
**Pull request by oc666**
fix bug #26514 and add memcached driver (#27979 in the feature tracker)

### CODE REVIEW

**Comment by oc666** [comment]
Hey
Is there any netbeans plugin to avoid those check-style errors & warnings?
Thanks

**Comment by realityking** [comment]
You can configure the net beans code style to closely match the one we mandate but there's no replacement of actually running phpcs.

Also could you please add a stub for unit testing of the two new classes. Thanks.

File: libraries/joomla/cache/storage/memcached.php [hunk 1]
```diff
@@ -0,0 +1,480 @@
...
+
+defined('JPATH_PLATFORM') or die;
+
+/**
+ * Memcached cache storage handler
+ *
+ * @package     Joomla.Platform
+ * @subpackage  Cache
+ * @see         http://php.net/manual/en/book.memcached.php
+ * @since       11.1
```
Comment (realityking) [inline libraries/joomla/cache/storage/memcached.php]: Please use 12.1 for all new classes/methods/members.

**Comment by elkuku** [comment]
As for Netbeans/CodeSniffer, have a look at: http://docs.joomla.org/Joomla_CodeSniffer#Netbeans

File: libraries/joomla/cache/storage/memcached.php (hunk 1 above)
Comment (oc666) [inline libraries/joomla/cache/storage/memcached.php]: How do I add commits to the pull request?

**Comment by realityking** [comment]
I only found 8 warnings

```

FILE: ...AMP/htdocs/joomla-platform/libraries/joomla/cache/storage/memcached.php
--------------------------------------------------------------------------------
FOUND 8 ERROR(S) AND 3 WARNING(S) AFFECTING 11 LINE(S)
--------------------------------------------------------------------------------
  50 | ERROR   | Expected "if (...)\n...{...}\n...else\n"; found "if
     |         | (...)\n...{...}\n...else "
  65 | ERROR   | Expected "if (...)\n...{...}\n...else\n"; found "if
     |         | (...)\n...{...}\n...else "
  74 | ERROR   | Please consider the /* */ style for comments that span over
     |         | multiple lines.
  76 | ERROR   | Please put a space between the // and the start of comment
     |         | text; found "//$servers    = (isset($params['servers'])) ?
     |         | $params['servers'] : array();"
  80 | ERROR   | Please consider a blank line preceding your comment
  81 | ERROR   | Expected "if (...)\n...{...}\n...else\n"; found "if (...)
     |         | ...{...} ...else "
  85 | ERROR   | Instanciating new classes without parameters does not require
     |         | brackets.
 207 | ERROR   | Please start your comment with a capital letter; found "//
     |         | prevent double writes, write only if it doesn't exist else
     |         | replace"
--------------------------------------------------------------------------------
```

**Comment by oc666** [comment]
Thanks, Rouven
Fixed.

**Comment by realityking** [comment]
Couple more things to fix:

```

FILE: ...P/htdocs/joomla-platform/libraries/joomla/session/storage/memcached.php
--------------------------------------------------------------------------------
FOUND 2 ERROR(S) AFFECTING 2 LINE(S)
--------------------------------------------------------------------------------
 107 | ERROR | Please put your comment on a separate line *preceding* your
     |       | code; found "// $this->_db->close();"
 214 | ERROR | Please start your comment with a capital letter; found "// set
     |       | prune period"
--------------------------------------------------------------------------------

FILE: ...AMP/htdocs/joomla-platform/libraries/joomla/cache/storage/memcached.php
--------------------------------------------------------------------------------
FOUND 1 ERROR(S) AND 3 WARNING(S) AFFECTING 4 LINE(S)
--------------------------------------------------------------------------------
 82 | ERROR   | Please consider a blank line preceding your comment
--------------------------------------------------------------------------------
```

File: libraries/joomla/session/storage/memcache.php [hunk 2]
```diff
@@ -62,22 +62,18 @@ public function __construct($options = array())
...
+		$this->_compress	= $config->get('memcache_compress', false)?MEMCACHE_COMPRESSED:false;
+		$this->_persistent	= $config->get('memcache_persist', true);
 
 		// This will be an array of loveliness
-		$this->_servers = (isset($params['servers'])) ? $params['servers'] : array();
+		// @todo: multiple servers
+		$this->_servers = array(
+			array(
+				'host' => $config->get('memcache_server_host', 'localhost'),
+				'port' => $config->get('memcache_server_port', 11211)
```
Comment (realityking) [inline libraries/joomla/session/storage/memcache.php:74]: I just noticed that, why are you hardcoding a server configuration?

Comment (oc666) [inline libraries/joomla/session/storage/memcache.php:74]: It is not hard coding. It tries to use memcache_server_host & memcache_server_port (from global config). If it's not set in the config, there is fallback to memcached default.

File: libraries/joomla/cache/storage/memcached.php (hunk 1 above)
Comment (realityking) [inline libraries/joomla/cache/storage/memcached.php]: Alright, Rouven just can't read ;)

**Comment by realityking** [comment]
Thanks!

**Pull request merged**